### PR TITLE
component: add possibility of executing single components on other cores

### PR DESCRIPTION
**Pull request by tlauda**
This set of patches adds possibility of executing single component on other core. Rest of the pipeline can run on different core.

### CODE REVIEW

File: src/include/sof/drivers/idc.h [hunk 1]
```diff
@@ -82,11 +82,21 @@
...
 #define iTS(x)	(((x) >> IDC_TYPE_SHIFT) & IDC_TYPE_MASK)
 
+/** \brief Max IDC message payload size in bytes. */
+#define IDC_MAX_PAYLOAD_SIZE	96
+
+/** \brief IDC message payload. */
+struct idc_payload {
+	uint8_t data[IDC_MAX_PAYLOAD_SIZE];
+};
+
```
Comment (paulstelian97) [inline src/include/sof/drivers/idc.h]: How do you ensure you never have two cores sending to the same target core at the same time, overwriting payloads?

File: src/audio/buffer.c [hunk 2]
```diff
@@ -40,6 +41,14 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
 		return NULL;
 	}
 
+	buffer->lock = rzalloc(SOF_MEM_ZONE_RUNTIME, SOF_MEM_FLAG_SHARED,
+			       SOF_MEM_CAPS_RAM, sizeof(*buffer->lock));
```
Comment (paulstelian97) [inline src/audio/buffer.c]: Lock being in a separate location from the buffer? This feels a bit sketchy, do we have a good reason to not move the buffer itself?

File: src/include/sof/audio/buffer.h [hunk 3]
```diff
@@ -67,6 +68,7 @@ struct comp_buffer {
 	uint32_t pipeline_id;
 	uint32_t caps;
 	uint32_t core;
+	bool inter_core;	/* true if between comps on different cores */
```
Comment (paulstelian97) [inline src/include/sof/audio/buffer.h]: I'd rephrase the comment a little: "True if connected to a component from another core", to make it a bit clearer.

File: src/include/sof/audio/buffer.h [hunk 4]
```diff
@@ -134,6 +134,31 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes);
...
+static inline void buffer_lock(struct comp_buffer *buffer)
+{
+	if (!buffer->inter_core)
+		return;
+
+	spin_lock(buffer->lock);
+
+	/* invalidate in case something has changed during our wait */
+	dcache_invalidate_region(buffer, sizeof(*buffer));
+}
```
Comment (paulstelian97) [inline src/include/sof/audio/buffer.h]: Ah, this commit explains the separate allocation of the spinlock. Got it.

File: src/audio/pipeline.c [hunk 5]
```diff
@@ -720,7 +727,7 @@ static int pipeline_comp_copy(struct comp_dev *current, void *data, int dir)
 	}
 
 	/* copy to downstream immediately */
-	if (dir == PPL_DIR_DOWNSTREAM) {
+	if (dir == PPL_DIR_DOWNSTREAM && cpu_is_me(current->comp.core)) {
```
Comment (paulstelian97) [inline src/audio/pipeline.c]: I feel this change can be separated, only running "copy" on the components on the current core but walking the entire pipeline?

File: src/include/sof/audio/buffer.h (hunk 3 above)
Comment (tlauda) [reply src/include/sof/audio/buffer.h]: Done.

File: src/audio/pipeline.c (hunk 5 above)
Comment (tlauda) [reply src/audio/pipeline.c]: Extracted to separate commit.

File: src/include/sof/drivers/idc.h (hunk 1 above)
Comment (tlauda) [reply src/include/sof/drivers/idc.h]: We don't. It something that probably should be done in the future, but right now we are working in master-slave mode, so only master core is sending messages to other cores.

Comment (paulstelian97) [reply src/include/sof/drivers/idc.h]: I'd say it's a high priority issue because you could potentially have weird arrangements with components, like pipeline 1 on core 0, pipeline 2 on core 1, and pipeline 1 has a component on core 1 and pipeline 2 has component on core 0. Topology is coming from userspace so we should be ready for this (even if by simply rejecting unsupported configurations initially).

**Comment by tlauda** [comment]
SOFCI TEST

File: src/include/sof/drivers/idc.h (hunk 1 above)
Comment (jajanusz) [reply src/include/sof/drivers/idc.h]: If I understand correctly @tlauda meant that messages are delivered from core 0 to other cores, not that it supports only pipelines on core 0 having components on other cores. If it's true then there is no issue atm.

Comment (paulstelian97) [reply src/include/sof/drivers/idc.h]: I thought from the pull request that messages will come from the actual core that owns the pipeline and not the core 0?

Comment (tlauda) [reply src/include/sof/drivers/idc.h]: @paulstelian97 Messages are always coming from master core, which is the only one receiving IPCs.

File: src/drivers/intel/cavs/idc.c [hunk 6]
```diff
@@ -217,6 +217,107 @@ static void idc_ipc(void)
...
+ * \brief Executes IDC component params message.
+ * \param[in] comp_id Component id to have params set.
+ * \return Error code.
+ */
+static int idc_params(uint32_t comp_id)
+{
+	struct ipc *ipc = ipc_get();
+	struct ipc_comp_dev *ipc_dev;
+	struct idc *idc = *idc_get();
+	struct idc_payload *payload = idc->payload + cpu_get_id();
```
Comment (jajanusz) [inline src/drivers/intel/cavs/idc.c]: Can we have func/macro for getting payload for given core?

**Comment by lgirdwood** [comment]
Wont this be 3 pipelines ? (pipeline being a set of components scheduled together in same slice/context) e.g.

```Pipe0 (core0) -> pipe1(core1) -> pipe2(core0)```

Since pipe2 has a scheduling dependency on pipe1,0 and pipe1 has dependency on pip0 ?

**Comment by tlauda** [comment]
SOFCI TEST

File: src/audio/pipeline.c [hunk 7]
```diff
@@ -285,26 +285,31 @@ static void pipeline_set_params(struct comp_dev *comp,
...
+	struct list_item *curr;
 	struct comp_buffer *buffer;
 	int i;
 
 	/* set comp params */
 	comp->direction = params->params.direction;
 
-	/* set buffer params */
-	buffer_list = comp_buffer_list(comp, dir);
+	while (clist != buffer_list) {
```
Comment (lyakh) [inline src/audio/pipeline.c]: `for (clist = buffer_list->next; clist != buffer_list; clist = clist->next)` eliminates the need for `curr`

Comment (tlauda) [reply src/audio/pipeline.c]: @lyakh Then you will access buffer after invalidating it. With my approach I can limit number of buffer locks in params and prepare calls of audio components to only the ones, that in fact change the format.

**Comment by lyakh** [comment]
@tlauda a general question to this PR: how is this going to play with virtualisation / #2376?  I admit I didn't look through the series, thoroughly, but does it mean, that the .copy() method for some components can run on a different core from the one, that's running `pipeline_copy()`? Actually, unrelated to this PR, we already now have an ability to run pipelines on different cores, right? So, it can happen that while streaming is taking place on a pipeline, running on core 1, a VM begins to boot and sends IPCs to create its own audio pipelines, some of which connect to the currently streaming pipeline. To protect against this we really need to spin-lock list manipulation and traversal?

File: src/audio/pipeline.c (hunk 7 above)
Comment (lyakh) [reply src/audio/pipeline.c]: @tlauda are you talking about invalidation, added in a separate patch? Which one is this?

**Comment by lgirdwood** [comment]
@tlauda ping on the pipelines question ? Do we actually define a single pipeline that spans > 1 core. How do we take care of scheduling dependencies ?

File: src/drivers/intel/cavs/idc.c (hunk 6 above)
Comment (tlauda) [reply src/drivers/intel/cavs/idc.c]: Done.

**Comment by tlauda** [comment]
> @tlauda ping on the pipelines question ? Do we actually define a single pipeline that spans > 1 core. How do we take care of scheduling dependencies ?

Initial idea was to define multiple pipelines, but it creates overhead by creating additional pipeline just to execute e.g. one component. After discussion with @mmaka1 and @lbetlej we've decided to allow running single component on slave cores. Overall it scales better. As for scheduling it works correctly for components with the similar frame size to the rest of the pipeline and it creates just one period of additional latency. To have it done perfectly we would need to switch such component's scheduling to EDF and calculate its own budget, but it's the task for the future.

File: src/audio/pipeline.c (hunk 7 above)
Comment (tlauda) [reply src/audio/pipeline.c]: @lyakh It's done in buffer_lock() function.

Comment (lyakh) [reply src/audio/pipeline.c]: @tlauda then sorry, I don't understand how replacing "while" with "for" would change anything about locking / cache invalidation. The only difference I see is that in your case the `.next` pointer is dereferenced before invalidating caches and in the existing / my case - after. But you anyway read in `buffer` into the cache in `buffer_set_params()` when checking `buffer->hw_params_configured`

File: src/include/sof/audio/buffer.h [hunk 8]
```diff
@@ -145,6 +145,31 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes);
...
+static inline void buffer_unlock(struct comp_buffer *buffer)
+{
+	if (!buffer->inter_core)
+		return;
+
+	/* save lock pointer to avoid memory access after cache flushing */
+	spinlock_t *lock = buffer->lock;
+
+	/* wtb and inv to avoid buffer locking in read only situations */
+	dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
```
Comment (lyakh) [inline src/include/sof/audio/buffer.h]: do you really need to invalidate the caches both in _lock and in _unlock?

File: src/audio/pipeline.c (hunk 7 above)
Comment (tlauda) [reply src/audio/pipeline.c]: @lyakh And after that I'm doing buffer_unlock(), which will do writeback and invalidate. Now from this point I don't want to load the same buffer again into the local cache, but use the fact that the memory is invalidated and automatically load new data (without doing explicit invalidation) in the places, where I'm just reading parameters without doing any modifications to the buffer. This kind of solution works OK with current FW flows, because buffer params are only modified during params and prepare calls.

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
SOFCI TEST

File: src/audio/pipeline.c (hunk 7 above)
Comment (lyakh) [reply src/audio/pipeline.c]: @tlauda Right, as long as you do invalidate caches in unlock, which I don't understand why it's necessary. And do you expect all (or at least a significant number of) buffers be in cache when you enter that function? If they aren't in cache, then the same thing happens anyway: a buffer is read into cache to get its `.next` pointer, then it's invalidated, then it's read in again and invalidated again.

Comment (tlauda) [reply src/audio/pipeline.c]: @lyakh As I stated in my previous comment - it's necessary in order to avoid calling buffer_lock/unlock in component_params/prepare calls, where buffer's params are only read, not modified. I think I see your point here and getting next pointer should be moved after buffer_lock.

File: src/audio/dai.c [hunk 9]
```diff
@@ -139,6 +139,8 @@ static struct comp_dev *dai_new(struct sof_ipc_comp *comp)
 	if (!dev)
 		return NULL;
 
+	dev->size = COMP_SIZE(struct sof_ipc_comp_dai);
```
Comment (lgirdwood) [inline src/audio/dai.c]: What about variable arrays in some of these IPCs ?

File: src/drivers/intel/cavs/idc.c [hunk 10]
```diff
@@ -254,14 +263,37 @@ static int idc_prepare(uint32_t comp_id)
...
+	struct comp_dev *dev;
 	int ret;
 
 	ipc_dev = ipc_get_comp_by_id(ipc, comp_id);
 	if (!ipc_dev)
 		return -ENODEV;
 
+	dev = ipc_dev->cd;
+
+	if (!dev->task) {
```
Comment (lgirdwood) [inline src/drivers/intel/cavs/idc.c]: Can you comment on this check.

File: src/include/sof/audio/buffer.h [hunk 11]
```diff
@@ -76,6 +77,7 @@ struct comp_buffer {
 	uint32_t pipeline_id;
 	uint32_t caps;
 	uint32_t core;
+	bool inter_core; /* true if connected to a comp from another core */
```
Comment (lgirdwood) [inline src/include/sof/audio/buffer.h]: Doe it save further logic if we have a flag for source and sink inter_core connections. I can see that using this flag will mean further repeated logic to check if its a source/sink connection ?

File: src/audio/component.c [hunk 12]
```diff
@@ -278,26 +281,40 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
...
+
+		buffer_unlock(buf);
 	} else {
 		/* for other components we iterate over all downstream buffers
 		 * (for playback) or upstream buffers (for capture).
 		 */
 		buffer_list = comp_buffer_list(dev, dir);
+		clist = buffer_list->next;
+
+		while (clist != buffer_list) {
```
Comment (lgirdwood) [inline src/audio/component.c]: can you comment on what we are looking for here

File: src/include/sof/audio/buffer.h (hunk 8 above)
Comment (lgirdwood) [reply src/include/sof/audio/buffer.h]: Can we also comment on what these function do/permit t the high level (at the start if each).

File: src/include/sof/audio/audio_stream.h [hunk 13]
```diff
@@ -220,6 +220,38 @@ static inline void audio_stream_init(struct audio_stream *buffer,
 	audio_stream_reset(buffer);
 }
 
+static inline void audio_stream_invalidate(struct audio_stream *buffer,
+					   uint32_t bytes)
+{
+	uint32_t head = bytes;
+	uint32_t tail = 0;
+
+	if ((char *)buffer->r_ptr + bytes > (char *)buffer->end_addr) {
```
Comment (lgirdwood) [inline src/include/sof/audio/audio_stream.h]: what are we checking for here ?

File: src/include/sof/audio/audio_stream.h [hunk 14]
```diff
@@ -220,6 +220,38 @@ static inline void audio_stream_init(struct audio_stream *buffer,
...
+		dcache_invalidate_region(buffer->addr, tail);
+}
+
+static inline void audio_stream_writeback(struct audio_stream *buffer,
+					  uint32_t bytes)
+{
+	uint32_t head = bytes;
+	uint32_t tail = 0;
+
+	if ((char *)buffer->w_ptr + bytes > (char *)buffer->end_addr) {
```
Comment (lgirdwood) [inline src/include/sof/audio/audio_stream.h]: and here

File: src/include/sof/audio/audio_stream.h [hunk 15]
```diff
@@ -220,6 +220,38 @@ static inline void audio_stream_init(struct audio_stream *buffer,
...
+		head = (char *)buffer->end_addr - (char *)buffer->w_ptr;
+		tail = bytes - head;
+	}
+
+	dcache_writeback_region(buffer->w_ptr, head);
+	if (tail)
+		dcache_writeback_region(buffer->addr, tail);
+}
```
Comment (lgirdwood) [inline src/include/sof/audio/audio_stream.h]: Btw, I would rename head/tail to include "size" as I initially read this code as some list iteration...

File: src/audio/asrc/asrc.c [hunk 16]
```diff
@@ -836,11 +831,17 @@ static int asrc_copy(struct comp_dev *dev)
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
+	buffer_lock(source);
+	buffer_lock(sink);
```
Comment (lgirdwood) [inline src/audio/asrc/asrc.c]: should we have an API here to guarantee ordering (and to avoid deadlocks).
i.e.
```c
buffer_lock_source_sink(source, sink);
...
buffer_unlock_source_sink(source, sink);
```
This may need to be atomic too.

File: src/audio/tone.c [hunk 17]
```diff
@@ -612,17 +618,22 @@ static int tone_copy(struct comp_dev *dev)
...
 
 	comp_dbg(dev, "tone_copy()");
 
 	/* tone component sink buffer */
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
+	buffer_lock(sink);
+	free = sink->stream.free;
+	buffer_unlock(sink);
```
Comment (lgirdwood) [inline src/audio/tone.c]: Another buffer_ API call ?
```
buffer_get_stream_free()
```
Probably a few other buffer calls could benefit here with locking ?

File: src/include/sof/audio/component.h [hunk 18]
```diff
@@ -617,7 +617,8 @@ static inline int comp_copy(struct comp_dev *dev)
 
 	assert(dev->drv->ops.copy);
 
-	ret = dev->drv->ops.copy(dev);
+	if (cpu_is_me(dev->comp.core))
```
Comment (lgirdwood) [inline src/include/sof/audio/component.h]: Can we have a comment here

File: src/ipc/ipc.c [hunk 19]
```diff
@@ -315,21 +316,29 @@ static int ipc_comp_to_buffer_connect(struct ipc_comp_dev *comp,
...
-		return -EINVAL;
-	}
-
 	if (!cpu_is_me(comp->core))
 		return ipc_process_on_core(comp->core);
 
 	trace_ipc("ipc: comp sink %d, source %d  -> connect", buffer->id,
 		  comp->id);
 
+	if (buffer->core != comp->core) {
```
Comment (lgirdwood) [inline src/ipc/ipc.c]: Need a comment here too

File: src/ipc/ipc.c [hunk 20]
```diff
@@ -315,21 +316,29 @@ static int ipc_comp_to_buffer_connect(struct ipc_comp_dev *comp,
...
+		dcache_invalidate_region(buffer->cb, sizeof(*buffer->cb));
+
+		buffer->cb->inter_core = true;
+
+		if (!comp->cd->is_shared) {
+			comp->cd = comp_make_shared(comp->cd);
+			if (!comp->cd)
+				return -ENOMEM;
```
Comment (lgirdwood) [inline src/ipc/ipc.c]: I guess this is fatal, should we trace_error it ?

File: src/ipc/ipc.c [hunk 21]
```diff
@@ -341,21 +350,29 @@ static int ipc_buffer_to_comp_connect(struct ipc_comp_dev *buffer,
...
-
-	if (!cpu_is_me(buffer->core))
-		return ipc_process_on_core(buffer->core);
+	if (!cpu_is_me(comp->core))
+		return ipc_process_on_core(comp->core);
 
 	trace_ipc("ipc: comp sink %d, source %d  -> connect", comp->id,
 		  buffer->id);
 
+	if (buffer->core != comp->core) {
```
Comment (lgirdwood) [inline src/ipc/ipc.c]: Needs another comment to explain

Comment (tlauda) [reply src/ipc/ipc.c]: Done.

File: src/ipc/ipc.c (hunk 19 above)
Comment (tlauda) [reply src/ipc/ipc.c]: Done.

File: src/ipc/ipc.c (hunk 20 above)
Comment (tlauda) [reply src/ipc/ipc.c]: Trace is inside of `comp_make_shared`.

File: src/include/sof/audio/component.h (hunk 18 above)
Comment (tlauda) [reply src/include/sof/audio/component.h]: Done.

File: src/include/sof/audio/audio_stream.h (hunk 15 above)
Comment (tlauda) [reply src/include/sof/audio/audio_stream.h]: Done.

File: src/include/sof/audio/audio_stream.h (hunk 13 above)
Comment (tlauda) [reply src/include/sof/audio/audio_stream.h]: Comment added.

File: src/include/sof/audio/audio_stream.h (hunk 14 above)
Comment (tlauda) [reply src/include/sof/audio/audio_stream.h]: Comment added.

File: src/drivers/intel/cavs/idc.c (hunk 10 above)
Comment (tlauda) [reply src/drivers/intel/cavs/idc.c]: Done.

File: src/include/sof/audio/component.h [hunk 22]
```diff
@@ -479,18 +487,24 @@ static inline int comp_cmd(struct comp_dev *dev, int cmd, void *data,
...
+		goto out;
 	}
 
 	if (dev->drv->ops.cmd)
-		return dev->drv->ops.cmd(dev, cmd, data, max_data_size);
-	return -EINVAL;
+		ret = dev->drv->ops.cmd(dev, cmd, data, max_data_size);
+
+out:
+	if (dev->is_shared)
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/include/sof/audio/component.h]: Done.

File: src/include/sof/audio/component.h [hunk 23]
```diff
@@ -296,6 +297,7 @@ struct comp_dev {
 				     */
 
 	uint32_t size;		/**< component's allocated size */
+	bool is_shared;		/**< indicates whether component is shared */
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/include/sof/audio/component.h]: Done.

File: src/audio/dai.c (hunk 9 above)
Comment (tlauda) [reply src/audio/dai.c]: `comp_dev` structures are allocated using size calculated from this macro, so I assume it's working OK.

File: src/audio/buffer.c [hunk 24]
```diff
@@ -151,12 +164,14 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes)
 	}
 
 	irq_local_disable(flags);
+	buffer_lock(buffer);
 
 	audio_stream_produce(&buffer->stream, bytes);
 
 	notifier_event(buffer, NOTIFIER_ID_BUFFER_PRODUCE,
 		       NOTIFIER_TARGET_CORE_LOCAL, &cb_data, sizeof(cb_data));
```
Comment (paulstelian97) [inline src/audio/buffer.c]: I hope there is no issue with only notifying locally.

File: src/include/sof/audio/buffer.h (hunk 11 above)
Comment (tlauda) [reply src/include/sof/audio/buffer.h]: Sink/source flags don't give us anything, since the same buffer is the source for the one of the components and sink for other.

File: src/audio/component.c (hunk 12 above)
Comment (tlauda) [reply src/audio/component.c]: I haven't added this loop, just modified it to avoid unnecessary buffer invalidation.

File: src/audio/asrc/asrc.c (hunk 16 above)
Comment (tlauda) [reply src/audio/asrc/asrc.c]: Ordering doesn't really matter. We just need to have it locked before getting buffer's current parameters like w_/r_ptr, free, avail etc. It just happens that both of these buffers are locked in the same place. I think adding additional functions is a bit misleading in this particular case. However we need to think in the near future about some kind of additional wrappers for getting/releasing buffers to avoid explicit lock/unlock calls.

File: src/audio/tone.c (hunk 17 above)
Comment (tlauda) [reply src/audio/tone.c]: Locking it on such granular level isn't efficient and will lead to deadlocks. As I stated in the previous comment, we need to think in the near future about some kind of additional wrappers for getting/releasing buffers to avoid explicit lock/unlock calls. Then the code will look much more cleaner and it will be easier for the integrators to implement their own processing components.

File: src/audio/buffer.c (hunk 24 above)
Comment (tlauda) [reply src/audio/buffer.c]: @paulstelian97 Probably it will need to be changed to multicore notification.

Comment (paulstelian97) [reply src/audio/buffer.c]: Can you remind me, who is actually listening on these produce/consume events?

Comment (tlauda) [reply src/audio/buffer.c]: Currently there are no clients for those calls, that's why I said "probably". But since there are no clients and no actual components running on slave cores I would say it's a low priority for now.

Comment (paulstelian97) [reply src/audio/buffer.c]: Alright, still this needs to be kept in mind for the future (it's a way to get weird bugs in the future, we may forget about this)

File: src/audio/asrc/asrc.c (hunk 16 above)
Comment (lgirdwood) [reply src/audio/asrc/asrc.c]: What happens if
```
lock(source);
IRQ -> lock(sink);->lock(source) -> DEADLOCK
```
or
```
lock(source);
lock(sink);
update(source);
IRQ -> update(source); ->done <<<< just trashed source
update(sink);
unlock(source);
unlock(sink);
```
We should really make this atomic.

Comment (tlauda) [reply src/audio/asrc/asrc.c]: @lgirdwood You are rather suggesting that we should also block irq, because we can get preempted, but still there is no correlation between locking source and sink buffers, since they are totally separate instances. As for preemption, we are setting highest irq lvl on the xtensa arch level when executing any interrupt, so technically it won't happen, but we can add this to make it more generic for other architectures in the future.

File: src/audio/tone.c (hunk 17 above)
Comment (paulstelian97) [reply src/audio/tone.c]: Maybe the wrapper copy() code could take the lock for both buffers in some atomic fashion (ordering may be sufficient, first source then sink) and then call the specific copy callback, which doesn't need to worry about this at all?

Comment (tlauda) [reply src/audio/tone.c]: @paulstelian97 That was my initial idea, but we can't block the whole copy. It's making the idea of multicore processing useless.

Comment (paulstelian97) [reply src/audio/tone.c]: Alright, but I don't think there is a single recipe which will work for all components either.

For my future use case, where I have processing functions that expect linear buffers (don't know how to interact with the SOF buffers and are closed source so I cannot update them), what would be the best recipe for locking and unlocking buffers? Would there be a need to allocate an additional buffer for that situation?

Comment (lgirdwood) [reply src/audio/tone.c]: I don't think we have to block the whole copy in this use case. The only locking and atomic code should be around R/W pointer updates and free/avail updates. Nothing else.

Comment (paulstelian97) [reply src/audio/tone.c]: So the buffer_update_*, buffer_consume_* and buffer_produce_* could have an internal lock? We have to assume there is only one producer and one consumer per buffer (it can be established and isn't a real limitation I guess).

Yup, this can work.

Comment (tlauda) [reply src/audio/tone.c]: @lgirdwood It's done exactly this way now. Would like to also explicitly block irq?

Comment (lgirdwood) [reply src/audio/tone.c]: @tlauda ok great, I wasn't sure from the PR diffs, yes I think the free/avail calc should be atomic (and block IRQs).

File: src/include/sof/audio/buffer.h (hunk 8 above)
Comment (lyakh) [reply src/include/sof/audio/buffer.h]: @tlauda could you please comment on this? Additionally, maybe this question was already asked and answered in one of discussions, marked as "resolved" and therefore closed above, sorry if that's the case. But I'm still unclear how about this this works: you start a pipeline processing on core 0, it processes components one by one, until it reaches component A, at which stage it requests core 1 to process that component. But what does core 0 do during that time? Just waits for core 1 to complete? Or processing asynchronous events like IPC? But usually there aren't many of those during streaming? What are we gaining?

Comment (tlauda) [reply src/include/sof/audio/buffer.h]: @lyakh I think I have already did in different comment. It invalidates to avoid calling buffer_lock/unlock in params and prepare calls, where nothing inside buffer is modified, but only read.
Core 0 only processes it's own components and doesn't wait for core 1. Processing on both cores happens totally asynchronous.

File: src/audio/tone.c (hunk 17 above)
Comment (tlauda) [reply src/audio/tone.c]: @lgirdwood Done.

**Comment by lgirdwood** [comment]
@tlauda jenkins looks good, but we have a UT failure on APL though.

**Comment by lgirdwood** [comment]
@tlauda Jenkins still is good - just looks like frequency drift between DUT and tester. Can you look into the internal CI fails.

**Comment by tlauda** [comment]
@lgirdwood It looks like some kind of regression on master. These fails are seen across all the PRs.

**Comment by lgirdwood** [comment]
@tlauda oh I see, @zrombel can this be bisected on internal CI ?

**Comment by lgirdwood** [comment]
@tlauda can you fix the conflict, I'd like to merge so we can validate the kernel. This wont stop the CI bisect.

**Comment by tlauda** [comment]
@lgirdwood Done.

File: src/include/sof/audio/buffer.h (hunk 8 above)
Comment (lyakh) [reply src/include/sof/audio/buffer.h]: @tlauda you're probably right, but sorry I still don't understand. I thought data processing in audio pipelines was linear: the next consumer can only start after the previous producer is done. I could understand if the other-core component was the first or the last in the pipeline, e.g.
```
core 0      core 1
A0
B0
A1            C0
B1
A2            C1
```
i.e. the original core can start processing the next data block while the "offload core" finishes processing the previous one. Is this what is used in this PR? I see that @lgirdwood is satisfied with this PR, so, it's probably just me who doesn't understand.

Comment (tlauda) [reply src/include/sof/audio/buffer.h]: @lyakh It's not always linear. We're going to skip components, which are not ours (owner is different core). This way other core can in the same time process the component inside our pipeline and with the next pipeline_copy we will have data to work on.

**Comment by tlauda** [comment]
SOFCI TEST

File: src/include/sof/audio/buffer.h (hunk 8 above)
Comment (paulstelian97) [reply src/include/sof/audio/buffer.h]: @lyakh Maybe there is some data from a _previous_ run of the pipeline that the next component can use. Sure, we'd have a bit of extra latency (1ms per component that is on a different core) but that may be acceptable.

So Host -> ParallelProcessing -> Dai:

First time Host would get the data and ParallelProcessing would be scheduled. DAI would have nothing to do.
Next millisecond Host would get new data, ParallelProcessing will have finished the previous ms and will be scheduled for the new data, DAI would get the old data from the previous ms.

**Comment by lgirdwood** [comment]
@dbaluta are we missing a PR for imx scripts, CI is failing on the iMX build atm ? Or are we just waiting on @xiulipan updating the container now ?
